### PR TITLE
Add thumbnails to dash

### LINKED ISSUE #3435: [Feature request] Thumbnails in DASH
**Is your feature request related to a problem? Please describe.**
I'm currently building a Roku frontend for Invidious ([Playlet](https://github.com/iBicha/playlet))
Roku expects thumbnails, aka storyboards (for "trick play" or scrubbing through the progress bar) to be included in the DASH-IF manifest ([DASH-standard thumbnail tiles](https://developer.roku.com/en-ca/docs/developer-program/media-playback/trick-mode/hls-and-dash.md#dash-standard-thumbnail-tiles))
So it would be cool to have an option where the storyboard included in the dash manifest, so that Roku can consume it properly

**Describe the solution you'd like**
`/api/manifest/dash/id/:videoid` would return a DASH manifest that has the storyboard info included ([like this example](https://developer.roku.com/en-ca/docs/developer-program/media-playback/trick-mode/hls-and-dash.md#dash-manifest-example))

**Describe alternatives you've considered**
I've tried to feed the storyboard manually into the system, and I did not succeed.
My options are to recreate the thumbnail data somehow (which is not trivial) or generate an extra `<AdaptationSet>` section for the thumbnails in the DASH manifest.
The latter is the most doable solution, but it's still a bad solution because
- I will have to serve the DASH manifest manually in a local server, proxying the request for the manifest, and injecting the thumbnails section
- I should not modify the DASH manifest on the fly like that, because I risk breaking it

**Additional context**
Is this feature is to be considered (I hope) maybe it could be an option like `/api/manifest/dash/id/:videoid?storyboard=true` so that  the feature remains backward compatible with existing API

### PR DESCRIPTION
**Pull request by iBicha**
This is an attempt to close #3435 

This allows seeking with a preview in Playlet
Here's what it looks like
It's not perfect, but it's a start

Note: I made sure that this is behind a flag (thumbnails) so that it does not change the behaviour for anyone who did not opt-in

thumbnails

### CODE REVIEW

**Comment by iBicha** [comment]
A little bump to draw the attention

**Comment by iBicha** [comment]
@unixfox @SamantazFox sorry for the ping, I'm not sure if this PR is in some sort of queue, or forgotten
Thanks

**Comment by SamantazFox** [comment]
@iBicha yeah, sorry, I haven't been much active this past month (the PR backlog have been filling up :c). I'll try to review and test that soon!

**Comment by iBicha** [comment]
> @iBicha yeah, sorry, I haven't been much active this past month (the PR backlog have been filling up :c). I'll try to review and test that soon!

Thank you, I know finding time for FOSS is hard! Cheers

File: src/invidious/routes/api/manifest.cr [hunk 1]
```diff
@@ -126,6 +127,62 @@ module Invidious::Routes::API::Manifest
...
+
+              storyboards.each do |storyboard|
+                interval = storyboard[:interval].to_i
+                # interval can be 0 https://github.com/iv-org/invidious/issues/3441
+                next if interval == 0
+
+                storyboard_height = storyboard[:storyboard_height]
+
+                # BUG: storyboardHeight can be wrong https://github.com/iv-org/invidious/issues/3440
+                # this fix/assumption is only if we have storyboard[:storyboard_count] = 1
```
Comment (iBicha) [inline src/invidious/routes/api/manifest.cr:145]: Based on discussion in https://github.com/iv-org/invidious/issues/3440, this is not the right assumption.
We need to "calculate" the image size based on which page we're on, storyboard count, and tile size

**Comment by iBicha** [comment]
@SamantazFox I'll refactor this after #4153 is merged

**Comment by iBicha** [comment]
Reflecting on this PR a bit, I will be closing this.
I'm trying to add DASH features primarily to support Playlet. I felt that the DASH generation belongs in Invidious, since it is a standard. But the more I dig into it, I discover limitations that are not part of the standard, and that are specific to Roku.
I asked Roku specifically about supporting tiles with different sizes (like in our case with the last page) and their response was in the lines of 
> ...There are all kinds of variations in DASH. Based on needs we could implement those but this needs to be based on a business-related need (like loss or revenue and so on).

So there will always be quirks or limitations, and they will not be prioritized unless they can justify it.

With that reasoning, I'll will be advocating to add things to Invidious that are not standards, or perhaps includes hacks and workarounds, so they fit the Roku OS. That does not sound right, that's why I'll be moving this piece to Playlet so I can cater to these cases (there are many differences between a standard dash player and Roku, including codecs, languages, containers and so on).

Sucks, but it is what it is.

Thank you!

**Pull request closed without merging**